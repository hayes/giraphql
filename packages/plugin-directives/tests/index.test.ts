import {
  printSchema,
  lexicographicSortSchema,
  GraphQLObjectType,
  GraphQLEnumType,
  GraphQLInputObjectType,
} from 'graphql';
import SchemaBuilder from '@giraphql/core';
import exampleSchema from './example/schema';

describe('extends example schema', () => {
  test('generates expected schema', () => {
    expect(printSchema(lexicographicSortSchema(exampleSchema))).toMatchSnapshot();
  });

  test('has expected directives in extensions', () => {
    const types = exampleSchema.getTypeMap();

    expect(types.Obj.extensions?.directives).toStrictEqual({ o: { foo: 123 } });
    expect(types.Query.extensions?.directives).toStrictEqual({
      o: { foo: 123 },
      rateLimit: { limit: 1, duration: 5 },
    });

    const testField = (types.Query as GraphQLObjectType).getFields().test;

    expect(testField.extensions?.directives).toStrictEqual({ f: [{ foo: 123 }] });
    expect(testField.args[0].extensions?.directives).toStrictEqual({ a: { foo: 123 } });

    expect(types.IF.extensions?.directives).toStrictEqual({ i: { foo: 123 } });
    expect(types.UN.extensions?.directives).toStrictEqual({ u: { foo: 123 } });
    expect(types.EN.extensions?.directives).toStrictEqual({ e: { foo: 123 } });
    expect((types.EN as GraphQLEnumType).getValue('ONE')?.extensions?.directives).toStrictEqual({
      ev: { foo: 123 },
    });
    expect(types.Date.extensions?.directives).toStrictEqual({ s: { foo: 123 } });
    expect(types.In.extensions?.directives).toStrictEqual({ io: { foo: 123 } });
    expect(
      (types.In as GraphQLInputObjectType).getFields().test.extensions?.directives,
    ).toStrictEqual({ if: { foo: 123 } });
  });

  test('gatsby format', () => {
    const builder = new SchemaBuilder<{
      Directives: {
        rateLimit: {
          locations: 'OBJECT' | 'FIELD_DEFINITION';
          args: {
            limit: number;
            duration: number;
          };
        };
      };
    }>({
      plugins: ['GiraphQLDirectives'],
    });

    builder.queryType({
      directives: {
        rateLimit: { limit: 5, duration: 60 },
      },
    });

    const types = builder.toSchema({}).getTypeMap();

    expect(types.Query.extensions?.directives).toStrictEqual([
      {
        name: 'rateLimit',
        args: { limit: 5, duration: 60 },
      },
    ]);
  });
});
