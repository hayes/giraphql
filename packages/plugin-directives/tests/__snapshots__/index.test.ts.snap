// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extends example schema generates expected schema 1`] = `
"scalar Date

enum EN {
  ONE
}

interface IF {
  field: String!
}

input In {
  test: String
}

type Obj {
  field: String!
}

type Query {
  test(arg1: String): String!
}

union UN = Obj
"
`;
